File: pg/abc266/f/namori.cpp
#include <bits/stdc++.h>
using namespace std;

class Namori {
  public:
    int V;
    vector<vector<int> > G, graph;
    vector<int> deg, loop;
    vector<bool> flag;
    Namori(int node_size) : V(node_size), G(V), deg(V, 0), flag(V, false) {}
    void add_edge(int u, int v) {
        G[u].push_back(v), G[v].push_back(u);
        deg[u]++, deg[v]++;
    }
    void dfs(int u) {
        loop.push_back(u);
        for (int v : G[u]) {
            if (flag[v]) continue;
            flag[v] = true;
            dfs(v);
        }
    }
    void build() {
        graph.resize(V);
        queue<int> que;
        // 葉の頂点を que に突っ込む
        for (int i = 0; i < V; i++) {
            if (deg[i] == 1) {
                que.push(i);
                flag[i] = true;
            }
        }
        // 葉からサイクル上の頂点に向かって登っていく
        while (!que.empty()) {
            int p = que.front();
            que.pop();
            for (int v : G[p]) {
                if (flag[v]) continue;
                deg[v]--;
                //サイクルから端
                graph[v].push_back(p);
                //端からサイクル
                graph[p].push_back(v);
                if (deg[v] > 1) continue;
                que.push(v);
                flag[v] = true;
            }
        }
        // サイクル上の頂点を loop に格納
        for (int i = 0; i < V; i++) {
            if (flag[i]) continue;
            flag[i] = true;
            dfs(i);
            break;
        }
    }
};

File: pg/abc266/f/namori.py
import sys
from collections import deque

sys.setrecursionlimit(10**9)


class Namori:
    def __init__(self, node_size):
        self.V = node_size
        self.G = [[] for _ in range(self.V)]
        self.deg = [0 for _ in range(self.V)]
        self.loop = []
        self.root = [-1 for _ in range(self.V)]
        self.flag = [False for _ in range(self.V)]

    def add_edge(self, u, v):
        self.G[u].append(v)
        self.G[v].append(u)
        self.deg[u] += 1
        self.deg[v] += 1

    def dfs(self, u):
        self.loop.append(u)
        for v in self.G[u]:
            if self.flag[v]: continue
            self.flag[v] = True
            self.dfs(v)

    def build(self):
        self.graph = [[] for _ in range(self.V)]
        self.que = deque()

        for i in range(self.V):
            if self.deg[i] == 1:
                self.que.append(i)
                self.flag[i] = True

        while self.que:
            p = self.que.popleft()
            for v in self.G[p]:
                if self.flag[v]: continue
                self.deg[v] -= 1
                self.graph[v].append(p)
                self.graph[p].append(v)
                if self.deg[v] > 1: continue
                self.que.append(v)
                self.flag[v] = True

        for i in range(self.V):
            if self.flag[i]: continue
            self.flag[i] = True
            self.dfs(i)
            break

        for r in self.loop:
            deq = deque([r])
            while deq:
                p = deq.popleft()
                if self.root[p] >= 0: continue
                self.root[p] = r
                for v in self.graph[p]:
                    if self.root[v] >= 0: continue
                    deq.append(v)
